Improve readability and backwards-compatibility of injection-code

betaboon opened the pull request.

mindflayer commented:
I like the function `_replace`, it makes the code cleaner. I am not sure I see the need for dividing the injecting code into sub-functions. I feet it's a bit of over-engineering with no real benefit. The module is clearly there already (`socket`/`ssl`/etc).
I will sleep on it, but I am not 100% sure I like this level of abstraction for something that simple as patching Python modules.

mindflayer commented:
Also, have a look at code coverage, not sure what happened but it went down.

betaboon commented:
sure. take your time.

i my mind having a dedicated place for each injection logic is more readable, especially considering the python- or urllib3-version specific things.

mindflayer commented:
That's the other thing, I don't think we need to add a new dependency for something as simple as trying to import something or not, I'd keep the `with contextlib.suppress(ImportError)` instead. It's more robust and easier to maintain.

betaboon commented:
sleep on it and ping me tomorrow and I'll take care of whatever decision you come to :)

mindflayer commented:
I've just had an idea. It would be nice if we had a dictionary of everything we patched, with the real and the mocked version, and use it to bring the status back, instead of using variables to know of I patched it or not that thing.

betaboon commented:
I'll give it a try

mindflayer commented:
It could be literally used both ways. As a way to define the monkey-patches and their counterparts.

betaboon commented:
ok.
i just pushed the following changes:
- removed `packaging`
- removed the individual functions from `inject.py`
- introduced something that might be close to your idea

mindflayer commented on mocket/inject.py:
```diff
@@ -27,99 +37,62 @@ def enable(
-
 
 def disable() -> None:
+    for module, name in list(_patches_restore.keys()):
```
I don't see the need for casting it to a `list`:

```suggestion
    for module, name in _patches_restore.keys():
```

betaboon replied:
as `_restore` removes elements from `_patch_restore` the keys have to be converted to list here as to prevent `RuntimeError: dictionary changed size during iteration`

mindflayer replied:
Makes sense, thank you!

mindflayer commented on mocket/utils.py:
```diff
@@ -6,12 +6,6 @@
 
 from mocket.compat import decode_from_bytes, encode_to_bytes
 
-# NOTE this is here for backwards-compat to keep old import-paths working
```
Why removing these import lines?

mindflayer replied:
Thank you :+1:

betaboon commented on mocket/inject.py:
```diff
@@ -1,24 +1,34 @@
+    original_value = _patches_restore.pop((module, name))
+    if original_value is None:
+        return
+    module.__dict__[name] = original_value
```
i personally don't like the walrus. but applied the change ;)

mindflayer replied:
I don't always use it, here I'd rather have a compact implementation than a nice-looking one. This said, I was more interested in reversing the logic of the `if` than adding the walrus operator. :)

mindflayer reviewed (approved):
LGTM!
